modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java

```diff
@@ -2644,78 +2651,128 @@ private synchronized void stop0(boolean cancel, boolean waitForBackups) {
             if (waitForBackups) {
                 waitingForBackups = true;
 
-                long curTopVer = grid.cluster().topologyVersion();
+                if (log.isInfoEnabled())
+                    log.info("Ensuring that caches have sufficient backups...");
 
-                Set<String> cacheNamesNoBackups;
+                boolean readyToStop = false;
 
-                if (log != null && log.isInfoEnabled())
-                    log.info("Ensuring that caches have sufficient backups...");
+                DistributedMetaStorage metaStorage = grid.context().distributedMetastorage();
 
-                boolean blocked;
+                while (!readyToStop && waitingForBackups) {
+                    boolean safeToStop = true;
 
-                do {
-                    blocked = false;
+                    long topVer = grid.cluster().topologyVersion();
 
-                    for (IgniteCacheProxy cache : grid().caches()) {
-                        GridCacheContext cctx = cache.context();
+                    HashSet<UUID> originalNodesToExclude;
 
-                        AffinityTopologyVersion readyTopVer = cctx.topology().readyTopologyVersion();
+                    HashSet<UUID> nodesToExclude;
 
-                        if (readyTopVer.topologyVersion() < curTopVer) {
-                            if (log != null) {
-                                LT.warn(log, "Delaying node shutdown due to ongoing topology change [curTopVer=" +
-                                    curTopVer + ", readyTopVer=" + readyTopVer + "]");
-                            }
+                    try {
+                        originalNodesToExclude = metaStorage.read(GRACEFUL_SHUTDOWN_METASTORE_KEY);
+
+                        nodesToExclude = originalNodesToExclude != null ? new HashSet<>(originalNodesToExclude) :
+                            new HashSet<>();
+                    }
+                    catch (IgniteCheckedException e) {
+                        U.error(log, "Unable to read " + GRACEFUL_SHUTDOWN_METASTORE_KEY +
+                            " value from metastore.", e);
 
-                            blocked = true;
+                        continue;
+                    }
 
-                            break;
-                        }
+                    for (CacheGroupContext grpCtx : grid.context().cache().cacheGroups()) {
```

Comment:
Should this code run on client nodes?
How will it work in case group context are not initialized (which is possible for clients)?

Reply:
The same for inactive clusters. How will this code work if we'll try to shutdown a server node safely on inactive cluster?

Reply:
1. Fixed;
2. I believe it'll work fine cause `grid.context().cache().cacheGroups()` will be empty, however in order to exclude unnecessary checks I've added `grid.cluster().active()` to `            if (waitForBackups && !grid.context().clientNode() && grid.cluster().active()) {...}`